File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque
input = sys.stdin.readline

n, m = map(int ,input().split())
g = [list(map(int, input().rstrip())) for _ in range(n)]
d = [(1, 0), (-1, 0), (0, 1), (0, -1)]
v = [[0]*m for _ in range(n)]

def in_range(y, x):
    if 0 <= y < n and 0 <= x < m:
        return True

def bfs(i, j):
    global v
    q = deque([(i, j)])
    v[i][j] = 1
    
    while q:
        a, b = q.popleft()

        for dy, dx in d:
            ny, nx = a + dy, b + dx
            if not in_range(ny, nx):
                continue
            if g[ny][nx] == 1 and v[ny][nx] == 0:
                q.append((ny, nx))
                v[ny][nx] = v[a][b] + 1

bfs(0, 0)
print(v[n-1][m-1])

File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const [first, ...input] = require('fs').readFileSync(process.platform == 'win32' ? "./input.txt" : '/dev/stdin').toString().trim().split("\n")

const [m, n] = first.trim().split(" ").map(Number);
const graph = input.map((line) => line.trim().split("").map(Number));

let queue = [];
dx = [-1, 1, 0, 0];
dy = [0, 0, -1, 1];

const bfs = (i, j) => {
    queue.push([i, j]);

    while (queue.length) {
        let [x, y] = queue.shift();

        for (let k = 0; k < 4; k++){
            let nx = x + dx[k];
            let ny = y + dy[k];
            
            if (0 <= nx && nx < m && 0 <= ny && ny < n) {
                if (graph[nx][ny] === 1) {
                    graph[nx][ny] = graph[x][y] + 1;
                    queue.push([nx, ny]);
                }
            }
        }
    }
};

const solution = () => { 
    bfs(0, 0);
    console.log(graph[m-1][n-1]);
}

solution();